aten/src/ATen/native/Histogram.cpp

```diff
@@ -0,0 +1,211 @@
+
+    TORCH_CHECK(bins.dim() == 1, "torch.histogram: bins tensor should have dimension 1,",
+            " but got ", bins.dim(), " dimension");
+
+    TORCH_CHECK(bins.numel() > 0, "torch.histogram: bins tensor should have at least 1 element,",
+            " but got ", bins.numel(), " elements");
+
+    if (weight.has_value()) {
+        TORCH_CHECK(input.dtype() == weight.value().dtype(), "torch.histogram: if weight tensor is provided,"
+                " input tensor and weight tensor should have the same dtype, but got input(", input.dtype(), ")",
+                ", and weight(", weight.value().dtype(), ")");
+
+        TORCH_CHECK(input.sizes() == weight.value().sizes(), "torch.histogram: if weight tensor is provided,"
+                " input tensor and weight tensor should have the same shape, but got input(", input.sizes(), ")",
+                ", and weight(", weight.value().sizes(), ")");
+    }
+}
+
+/* Checks properties of output tensors hist and bin_edges, then resizes them.
+ */
+void histogram_prepare_out(const Tensor& input, int64_t bin_ct,
+        const Tensor& hist, const Tensor& bin_edges) {
+    TORCH_CHECK(input.dtype() == hist.dtype(), "torch.histogram: input tensor and hist tensor should",
+            " have the same dtype, but got input ", input.dtype(), " and hist ", hist.dtype());
+
+    TORCH_CHECK(input.dtype() == bin_edges.dtype(), "torch.histogram: input tensor and bin_edges tensor should",
+            " have the same dtype, but got input ", input.dtype(), " and bin_edges ", bin_edges.dtype());
+
+    TORCH_CHECK(bin_ct > 0,
+            "torch.histogram(): bins must be > 0, but got ", bin_ct);
+
+    at::native::resize_output(hist, bin_ct);
+
+    at::native::resize_output(bin_edges, bin_ct + 1);
+
+    TORCH_CHECK(hist.is_contiguous(), "torch.histogram: hist tensor must be contiguous");
+}
+
+/* Determines the outermost bin edges.
+ */
+std::pair<double, double> select_outer_bin_edges(const Tensor& input,
+        const c10::optional<Scalar>& min, const c10::optional<Scalar>& max) {
+    if (min.has_value() && max.has_value()) {
+        return std::make_pair(min.value().to<double>(), max.value().to<double>());
+    }
+
+    // Default range for empty input matching numpy.histogram's default
+    if (input.numel() == 0) {
+        return std::make_pair(0., 1.);
+    }
+
+    auto extrema = _aminmax(input);
+
+    double leftmost_edge  = min.has_value() ? min.value().to<double>() : std::get<0>(extrema).item<double>();
+    double rightmost_edge = max.has_value() ? max.value().to<double>() : std::get<1>(extrema).item<double>();
+
+    TORCH_CHECK(leftmost_edge <= rightmost_edge, "torch.histogram: min should not exceed max, but got",
+            "min ", leftmost_edge, " max ", rightmost_edge);
+
+    // Expand empty range to match numpy behavior and avoid division by 0 in normalization
+    if (leftmost_edge == rightmost_edge) {
```

Comment:
Is this what NumPy does if min==max? Maybe this case should be an error?

Reply:
Bigger question: what does NumPy do if given a tensor with all the same values and no explicit range?

Reply:
This is the same as what NumPy does: https://github.com/numpy/numpy/blob/main/numpy/lib/histograms.py#L327

` numpy.histogram([0])`, for example, produces bins spanning the range [-0.5, 0.5].

I think it's reasonable to produce some form of output reflecting the information "all elements of input are equal" and that we shouldn't error out here.

Reply:
OK; sounds good